src/mpid/ch4/netmod/ofi/ofi_send.h

```diff
@@ -581,28 +521,92 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_send(const void *buf, MPI_Aint count, MPI
     void *send_buf = MPIR_get_contig_ptr(buf, dt_true_lb);
     MPIR_GPU_query_pointer_attr(send_buf, &attr);
 
-    if (MPIR_CVAR_CH4_OFI_ENABLE_INJECT && !syncflag && dt_contig &&
-        (data_sz <= MPIDI_OFI_global.max_buffered_send)) {
-        void *pack_buf = NULL;
-        if (data_sz && MPL_gpu_query_pointer_is_dev(send_buf, &attr)) {
-            MPIDI_OFI_register_am_bufs();
-            if (!MPIDI_OFI_ENABLE_HMEM) {
-                /* Force pack for GPU buffer. */
-                pack_buf = MPL_malloc(data_sz, MPL_MEM_OTHER);
-                mpi_errno = MPIR_Localcopy_gpu(send_buf, data_sz, MPI_BYTE, 0, &attr, pack_buf,
-                                               data_sz, MPI_BYTE, 0, NULL,
-                                               MPL_GPU_COPY_DIRECTION_NONE,
-                                               MPIDI_OFI_gpu_get_send_engine_type(), true);
-                MPIR_ERR_CHECK(mpi_errno);
-                send_buf = pack_buf;
+    bool is_gpu = false;
+    bool need_pack = false;
+    bool need_mr = false;
+    bool do_inject = false;
+    bool do_iov = false;
+    bool do_gpu_pipelining = false;
+    bool do_striping = false;
+    bool do_huge = false;
+
+    /* check gpu */
+    if (data_sz && MPL_gpu_query_pointer_is_dev(send_buf, &attr)) {
+        is_gpu = true;
+        MPIDI_OFI_register_am_bufs();
+    }
+
+    /* check gpu paths */
+    if (is_gpu) {
+        if (!MPIDI_OFI_ENABLE_HMEM) {
+            need_pack = true;
+        } else {
+            /* contig + ENABLE_HMEM */
+            if (dt_contig && MPIDI_OFI_ENABLE_MR_HMEM) {
+                if (data_sz >= MPIR_CVAR_CH4_OFI_GPU_RDMA_THRESHOLD &&
+                    attr.type == MPL_GPU_POINTER_DEV) {
+                    need_mr = true;
+                } else {
+                    need_pack = true;
+                }
             } else {
-                mpi_errno =
-                    MPIDI_OFI_send_normal(buf, count, datatype, cq_data, dst_rank, tag, comm,
-                                          context_offset, addr, vci_src, vci_dst, request,
-                                          dt_contig, data_sz, dt_ptr, dt_true_lb, syncflag, attr);
-                goto fn_exit;
+                /* FIXME: need cleanup. */
+                need_pack = true;
```

Comment:
I need help to get the gpu branching logic. The key question (that I am quite confused right now) is -
* What is the condition that we can use GPU RDMA direct?

Reply:
Condition is:

* `MPIDI_OFI_ENABLE_HMEM` (provider accepts GPU buffers)
* `dt_contig` (buffer is contiguous)
* `data_sz >= MPIR_CVAR_CH4_OFI_GPU_RDMA_THRESHOLD` (size meets the threshold)
* `attr.type == MPL_GPU_POINTER_DEV` (buffer is strictly a device pointer)

Reply:
From my understanding, the provider can only do GPU RDMA with GPU buffers, not with host-registered or shared buffers. That is the reason for the last point.

Also, `MPIDI_OFI_ENABLE_MR_HMEM` is a setting from the provider that indicates the buffer must be registered prior to passing to OFI. Just to keep that in mind

Reply:
Thanks @abrooks98 !  What about `MPIDI_OFI_ENABLE_MR_HMEM` check?

Reply:
So `MPIDI_OFI_ENABLE_MR_HMEM` does need to be checked, and if it is true, then the memory must be registered (i.e. via `MPIDI_OFI_register_memory_and_bind`) https://github.com/pmodels/mpich/blob/main/src/mpid/ch4/netmod/ofi/ofi_send.h#L267-L272

Reply:
Ok, that clarifies it.

What about shared ZE pointer (non-strict gpu buffer)?

Reply:
For shared ZE pointer that is not strictly a GPU buffer, it cannot be passed directly to OFI because it is still a GPU object. In this case we take the path of packing to a host buffer prior to passing to OFI

Reply:
So non-strict GPU buffer always needs packing? That's a simple check.

Reply:
@abrooks98 Also to confirm - that data_sz threshold is a performance concern due to `MR_HMEM` overhead, right?

Reply:
(good lord the existing code is not straightforward... glad you are refactoring this)

I think my previous comment is incorrect if I'm reading this correctly:

https://github.com/pmodels/mpich/blob/main/src/mpid/ch4/netmod/ofi/ofi_send.h#L256-L265

Only need to pack non-strict GPU buffers if either:

* MPIDI_OFI_ENABLE_HMEM is not true (OFI wouldn't accept any type of GPU buffer)
* buffer is not contiguous (must be contiguous in any scenario)
* MPIDI_OFI_ENABLE_MR_HMEM is true (OFI needs GPU buffers to be registered, and the non-strict GPU buffers are still GPU objects)

Reply:
So the case where `MPIDI_OFI_ENABLE_HMEM && dt_contig && !MPIDI_OFI_ENABLE_MR_HMEM` is when packing is not required for non-strict GPU buffers

Reply:
> @abrooks98 Also to confirm - that data_sz threshold is a performance concern due to `MR_HMEM` overhead, right?

correct